### PR TITLE
LPS-124021 Use mouseover/mouseout instead of mouseenter/mouseleave

### PR DESCRIPTION
**Pull request by julien**
With metal's `dom.delegate`, `mouseenter` and `mouseleave` events are
[mapped to `mouseover` and `mouseout`](https://github.com/metal/metal.js/blob/5b5032dd840bc3364894776c8bb2fb82517e2553/packages/metal-dom/src/events.js#L11-L34), so applying the same logic here 
for tooltips which fixed the issue.

Additionally, two new test cases were added for the `delegate` function.

### CODE REVIEW

**Comment by julien** [comment]
ci:test:sf

**Comment by julien** [comment]
ci:test:relevant

File: modules/apps/frontend-js/frontend-js-web/test/liferay/delegate/delegate.es.js [hunk 1]
```diff
@@ -181,4 +181,61 @@ describe('delegate', () => {
 
 		expect(delegateTarget).toBe(document.querySelector('.match'));
 	});
+
+	it('works with multiple selectors', () => {
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/delegate/delegate.es.js:185]: Just for the sake of precision, can we reword this to?

```js
it('works with comma-delimited selector lists', () => {
```

(because "selector list" is the technical name for this thing)

File: modules/apps/frontend-js/frontend-js-web/test/liferay/delegate/delegate.es.js [hunk 2]
```diff
@@ -181,4 +181,61 @@ describe('delegate', () => {
...
+
+		expect(listener).toHaveBeenCalledTimes(1);
+
+		element = getByTestId(document, 'two');
+		element.dispatchEvent(event);
+
+		expect(listener).toHaveBeenCalledTimes(2);
+	});
+
+	it('delegateTarget returns the correct selector', () => {
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/delegate/delegate.es.js:215]: Could this be made more specific too?

We already have `it('gives access to delegateTarget inside the callback'` higher up. What's the specific thing you're trying to check here? (Also, nit: note that it's not `return`-ing anything, and it's an element, not a selector.)

Ideally, whatever you come up with should read like a valid English sentence when read out loud (unlike "it delegateTarget returns"). Maybe something like:

```js
it('sets delegateTarget correctly when using a selector list', () => {
```

File: modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.js [hunk 3]
```diff
@@ -174,14 +174,14 @@ const TooltipProvider = () => {
...
-			'mouseenter',
+			'mouseover',
 			SELECTOR_TOOLTIP,
 			() => dispatch({target: state.target, type: 'show'})
 		);
 
 		const TOOLTIP_LEAVE = delegate(
 			document.body,
-			'mouseleave',
+			'mouseout',
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.js:184]: Slightly scary, but I think it makes sense. `mouseover` will fire on entering any child of `document.body`, whereas `mouseenter` would oly fire for `document.body` itself.

**Review by wincent — COMMENTED** [review]
Fix lg2m, but I have some pedantic requests for the title of the `it` blocks in the tests.

File: modules/apps/frontend-js/frontend-js-web/test/liferay/delegate/delegate.es.js (hunk 1 above)
Comment (julien) [reply modules/apps/frontend-js/frontend-js-web/test/liferay/delegate/delegate.es.js:185]: of course

File: modules/apps/frontend-js/frontend-js-web/test/liferay/delegate/delegate.es.js (hunk 2 above)
Comment (julien) [reply modules/apps/frontend-js/frontend-js-web/test/liferay/delegate/delegate.es.js:215]: Yes of course. What I'm testing here is that the `delegateTarget` returns the element that was clicked on. I could have modified the existing test.

File: modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.js (hunk 3 above)
Comment (julien) [reply modules/apps/frontend-js/frontend-js-tooltip-support-web/src/main/resources/META-INF/resources/index.js:184]: That's exactly what I saw when debugging, as you said `mouseenter` will fire only when entering  `document.body` and not it's children.

**Comment by jbalsas** [comment]
❤️

**Comment by jbalsas** [comment]
ci:test:sf

**Comment by jbalsas** [comment]
@julien, you'll have to re-run `ci:test-relevant` because you pushed additional commits. I'd suggest just merging the commits to have just the one (if @wincent is satisfied with the changes) and then trigger a `ci:stop` and try `ci:test:relevant` (plus `ci:sf`) again.

**Comment by julien** [comment]
@jbalsas makes sense and works for me too, thanks. I thought I'd squash into one commit but I also didn't want to "force" push until I had @wincent 's approval because it makes it hard to see all the things that happened.

**Comment by julien** [comment]
ci:stop

**Comment by julien** [comment]
Doing "ci:stop" now because I am going to squash anyway.

**Review by wincent — APPROVED** [review]
lg2m...

As for force-pushing vs not force-pushing: doesn't make much difference when the change is small (ie. I can look at the whole thing again without having to hunt through it to spot the differences).

**Comment by julien** [comment]
ci:test:relevant

**Comment by jbalsas** [comment]
@julien, while you're at this, could you check to see if these are also related and fixed here or if there's more to it?
- [Cannot hide tooltip when clicking blank space or click again](https://issues.liferay.com/browse/LPS-124219)
- [Duplicated tooltips and 'true' value into tooltip](https://issues.liferay.com/browse/LPS-124157)

**Comment by jbalsas** [comment]
Oh, I see you already commented there... if you think they're solved with this, just close it as duplicate linking to the other one and ask (or do it yourself) that the reproduction steps are also copied in the main one to make sure we keep that clean-ish

**Comment by julien** [comment]
Yeah sorry @jbalsas I did comment but didn't update anything - should be OK now.

I've also assigned the [Duplicated tooltips and 'true' value into tooltip](https://issues.liferay.com/browse/LPS-124157) issue to myself, but in this case it's a different issue.

**Comment by wincent** [comment]
> I've also assigned the [Duplicated tooltips and 'true' value into tooltip](https://issues.liferay.com/browse/LPS-124157) issue to myself, but in this case it's a different issue.

true

**Comment by jbalsas** [comment]
ci:forward
